src/frontend/src/expr/type_inference/func.rs

```diff
@@ -585,7 +585,7 @@ fn infer_type_for_special(
             }
             Ok(Some(DataType::Varchar))
         }
-        ExprType::ArrayContains | ExprType::ArrayContained => {
+        ExprType::ArrayContains | ExprType::ArrayContained | ExprType::ArrayOverlap => {
```

Comment:
We shall also ensure both sides of `ArrayIntersect` are of the same array type - either at the first dimension or flatten recursively (see comments above).